### PR TITLE
[libc++][iterator][ranges] P2997R1: Removing the common reference requirement from the indirectly invocable concepts

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Implements as DR against C++20: https://wg21.link/P2997R1

References:
- https://eel.is/c++draft/indirectcallable.indirectinvocable
- https://eel.is/c++draft/version.syn#header:%3cversion%3e

### CODE REVIEW

File: libcxx/include/version [hunk 1]
```diff
@@ -534,6 +535,8 @@ __cpp_lib_void_t                                        201411L <type_traits>
 # undef  __cpp_lib_out_ptr
 // # define __cpp_lib_out_ptr                              202311L
 // # define __cpp_lib_philox_engine                        202406L
+# undef  __cpp_lib_ranges
+# define __cpp_lib_ranges                               202406L
```
Comment (frederick-vs-ja) [inline libcxx/include/version:539]: It's weird to bump `__cpp_lib_ranges` to this value when [P2609R3](https://wg21.link/p2609r3) (which bumps `__cpp_lib_ranges` to `202302L`) hasn't been implemented.

Coincidentally (may be not), P2609R3 and P2997R1 change exactly the same set of concepts.

Comment (H-G-Hristov) [reply libcxx/include/version]: Good catch! Thanks!

File: libcxx/utils/generate_feature_test_macro_components.py [hunk 2]
```diff
@@ -993,7 +993,8 @@ def add_version_header(tc):
             "name": "__cpp_lib_ranges",
             "values": {
                 "c++20": 202207,
-                # "c++26": 202406, # P2997R1 Removing the common reference requirement from the indirectly invocable concepts
+                # "c++23": 202302,  # Relaxing Ranges Just A Smidge
+                # "c++26": 202406,  # P2997R1 Removing the common reference requirement from the indirectly invocable concepts
```
Comment (ldionne) [inline libcxx/utils/generate_feature_test_macro_components.py:997]:
```suggestion
                # "c++26": 202406,  # P2997R1 Removing the common reference requirement from the indirectly invocable concepts (already implemented as a DR)
```

File: libcxx/docs/ReleaseNotes/19.rst [hunk 3]
```diff
@@ -47,6 +47,7 @@ Implemented Papers
 - P3142R0 - Printing Blank Lines with ``println`` (as DR against C++23)
 - P2944R3 - Comparisons for ``reference_wrapper`` (comparison operators for ``reference_wrapper`` only)
 - P2968R2 - Make ``std::ignore`` a first-class object
+- P2997R1 - Removing the common reference requirement from the indirectly invocable concepts
```
Comment (ldionne) [inline libcxx/docs/ReleaseNotes/19.rst:50]: Please also update the status page.

**Review by ldionne — COMMENTED** [review]
@jwakely @StephanTLavavej Did you folks also implement this as a DR? It certainly makes sense to me -- I'm generally against extensions but having a different behavior in C++20 / C++26 seems really confusing for no reason. WDYT?

**Comment by Zingam** [comment]
> @jwakely @StephanTLavavej Did you folks also implement this as a DR? It certainly makes sense to me -- I'm generally against extensions but having a different behavior in C++20 / C++26 seems really confusing for no reason. WDYT?

I believe @frederick-vs-ja implemented P2997R1 in MS STL. 
cppreference listed this paper as DR. I am still not sure when papers are voted in as DRs:
https://en.cppreference.com/w/cpp/compiler_support/26

**Comment by jwakely** [comment]
> I am still not sure when papers are voted in as DRs:

When it says so in the straw poll.

But in some cases, like this one, implementers agree to do it despite it not being voted that way.

@ldionne I haven't thought about it yet, but we'll follow the herd if libc++ and MSVC do it.

File: libcxx/test/std/iterators/iterator.requirements/indirectcallable/indirectinvocable/indirectly_unary_invocable.compile.pass.cpp [hunk 4]
```diff
@@ -56,12 +56,12 @@ struct BadInvocable3 {
 };
 static_assert(!std::indirectly_unary_invocable<BadInvocable3, It>);
 
-// Should fail when the invocable can't be called with (iter_common_reference_t)
+// This case was made valid by P2997R1.
 struct BadInvocable4 {
```
[the remark this answers is not in the document]
Comment (H-G-Hristov) [reply libcxx/test/std/iterators/iterator.requirements/indirectcallable/indirectinvocable/indirectly_unary_invocable.compile.pass.cpp]: I hope you like the new name :)

File: libcxx/docs/ReleaseNotes/19.rst (hunk 3 above)
Comment (H-G-Hristov) [reply libcxx/docs/ReleaseNotes/19.rst]: Done! Thanks!

**Comment by StephanTLavavej** [comment]
Yep, we implemented this one as a de facto DR.

**Review by ldionne — APPROVED** [review]
LGTM, thanks.

@jwakely Per discussion in the thread, libc++ will do this one as a DR against C++20.

**Pull request merged**